AI/VCAI/Pathfinding/AINodeStorage.cpp

```diff
@@ -10,15 +10,61 @@
 #include "StdInc.h"
 #include "AINodeStorage.h"
 
+#include "../../../lib/PathfinderUtil.h"
+#include "../../../lib/mapping/CMap.h"
+#include "../../../lib/CPlayerState.h"
+
 
 AINodeStorage::AINodeStorage(const int3 & Sizes)
 	: sizes(Sizes)
 {
 	nodes.resize(boost::extents[sizes.x][sizes.y][sizes.z][EPathfindingLayer::NUM_LAYERS][NUM_CHAINS]);
 }
 
-AINodeStorage::~AINodeStorage()
+AINodeStorage::~AINodeStorage() = default;
+
+void AINodeStorage::initialize(const PathfinderOptions & options, const CGameState * gs, const CGHeroInstance * hero)
 {
+	//TODO: fix this code duplication with NodeStorage::initialize, problem is to keep `resetTile` inline
+
+	int3 pos;
+	int3 sizes = gs->getMapSize();
+	const auto & fow = static_cast<const CGameInfoCallback *>(gs)->getPlayerTeam(hero->tempOwner)->fogOfWarMap;
+	const PlayerColor player = hero->tempOwner;
+
+	//make 200% sure that these are loop invariants (also a bit shorter code), let compiler do the rest(loop unswitching)
+	const bool useFlying = options.useFlying;
+	const bool useWaterWalking = options.useWaterWalking;
+
+	for(pos.x=0; pos.x < sizes.x; ++pos.x)
```

Comment:
will be nice to overload some operators in int3 and replace 3 loops with something like that `for (int3 pos; pos<sizes; ++pos)`

Reply:
I have no idea how to make such custom iterator fast, you have to add not trivial logic to `++` and compare all 3 coordinates each iteration.